File: 142_detect_cycle.py
# 19'
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
	def detectCycle(self, head: ListNode) -> ListNode:
		# init
		map = []
		res = None
		pos = 0

		if head:
			if head.next:
				node = head
			else:
				return res
		else:
			return res

		while node:
			if node in map:
				res = node
				return res
			else:
				if node.next:
					map.append(node)
					node = node.next
				
		return res

File: 142_detect_cycle.java
// Version 0
// TC:O(N), SC:O(N)
// Hashset
/**
 * class ListNode{
 * 	int val;
 *  ListNode next;
 *  ListNode(int x){
 *  	val = x;
 *  	next = null;
 *  }	
 * }
 */

public class Solution{
	public ListNode detectCycle(ListNode head){
		ListNode pos = head;
		Set<ListNode> visited = new HashSet<>();
		while(pos != null){
			if(visited.contains(pos)){
				return pos;
			}else{
				visited.add(pos);
			}
			pos = pos.next;
		}
		return null;
	}
}


// Version 1
// double pointer
// TC: O(N), SC: O(1)
public class Solution{
	public ListNode detectCycle(ListNode head){
		if(head==null || head.next==null){
			return null;
		}
		// init
		ListNode slow = head, fast = head;

		// traverse 1st
		while(fast!=null){
			slow = slow.next;
			if(fast.next==null){
				return null;
			}else{
				fast = fast.next.next;
			}
			if(fast==slow){		// 2nd traverse, fast move 1 step this round
				fast = head;
				while(fast!=slow){
					fast = fast.next;
					slow = slow.next;
				}
				return fast;
			}
		}
		return null;
	}
}